pkg/instance/service.go

```diff
@@ -76,43 +80,267 @@ func (s service) SaveInstance(instance *model.DeploymentInstance) error {
+func (s service) validateNoCycles(instances []*model.DeploymentInstance) (graph.Graph[string, *model.DeploymentInstance], error) {
+	g := graph.New(func(instance *model.DeploymentInstance) string {
+		return instance.StackName
+	}, graph.Directed(), graph.PreventCycles())
+
+	for _, instance := range instances {
+		err := g.AddVertex(instance)
+		if err != nil {
+			return nil, fmt.Errorf("failed adding vertex for instance %q: %v", instance.Name, err)
+		}
+	}
+
+	for _, src := range instances {
+		stack, err := s.stackService.Find(src.StackName)
+		if err != nil {
+			return nil, err
+		}
+		for _, requiredStack := range stack.Requires {
+			err := g.AddEdge(src.StackName, requiredStack.Name)
+			if err != nil {
+				if errors.Is(err, graph.ErrEdgeAlreadyExists) {
+					return nil, fmt.Errorf("instance %q requires %q more than once", src.Name, requiredStack.Name)
+				} else if errors.Is(err, graph.ErrEdgeCreatesCycle) {
+					return nil, fmt.Errorf("edge from instance %q to instance %q creates a cycle", src.Name, requiredStack.Name)
+				}
+				return nil, fmt.Errorf("failed adding edge from instance %q to instance %q: %v", src.Name, requiredStack.Name, err)
+			}
+		}
+	}
+
+	return g, nil
+}
+
+func (s service) resolveParameters(deployment *model.Deployment) error {
+	for _, instance := range deployment.Instances {
+		instanceParameters := instance.Parameters
+
+		stack, err := s.stackService.Find(instance.StackName)
+		if err != nil {
+			return err
+		}
+
+		err = rejectConsumedParameters(instanceParameters, stack)
+		if err != nil {
+			return err
+		}
+
+		err = rejectNonExistingParameters(instanceParameters, stack)
+		if err != nil {
+			return err
+		}
+
+		addUnsuppliedParameters(stack, instanceParameters)
+
+		for name, parameter := range instanceParameters {
+			if !stack.Parameters[name].Consumed {
+				continue
+			}
+
+			for _, requiredStack := range stack.Requires {
+				// consume from instance parameters
+				sourceInstance := deployment.FindInstanceByStackName(requiredStack.Name)
+				if sourceInstance == nil {
+					return errdef.NewNotFound("failed to find required instance %q of instance %q", sourceInstance.Name, instance.Name)
+				}
+
+				if sourceInstanceParameter, ok := sourceInstance.Parameters[name]; ok {
+					parameter.Value = sourceInstanceParameter.Value
+				}
+
+				// consume from provider
+				if provider, ok := requiredStack.ParameterProviders[name]; ok {
+					value, err := provider.Provide(*sourceInstance)
+					if err != nil {
+						return fmt.Errorf("failed to provide value for instance %q parameter %q: %v", instance.Name, name, err)
+					}
+					parameter.Value = value
+				}
+
+				instanceParameters[name] = parameter
+			}
+		}
+	}
+
+	return nil
+}
+
+func rejectConsumedParameters(instanceParameters model.DeploymentInstanceParameters, stack *model.Stack) error {
+	var errs []error
+	for name := range instanceParameters {
+		if stack.Parameters[name].Consumed {
```

Comment:
Would this not panic given an instance parameter that is not part of the stack.Parameters 🤔 (since `rejectConsumedParameters` is called before `rejectNonExistingParameters`)